gen/appdb

```diff
@@ -38,88 +38,104 @@ my $vosStruc = {};
-			my @dn = ();
-			foreach my $subjectDN (keys %{$memberAttributes{$A_USER_CERT_DNS}}) {
-				my $subjectDNWithoutPrefix = $subjectDN;
-				$subjectDNWithoutPrefix =~ s/^[0-9]+[:]//;
-				push @dn, { "content" => $subjectDNWithoutPrefix,
-				            "ca" => $memberAttributes{$A_USER_CERT_DNS}->{$subjectDN},
-				          };
-			}
+		# prepare DNs
+		my $userCertDNs = $data->getUserAttributeValue(member => $memberId, attrName => $A_USER_CERT_DNS);
 
-			my $userName = "";
-			if($memberAttributes{$A_USER_FIRST_NAME}) {
-				$userName = $memberAttributes{$A_USER_FIRST_NAME};
-				$userName.= " " . $memberAttributes{$A_USER_LAST_NAME} if $memberAttributes{$A_USER_LAST_NAME};
-			} else {
-				$userName.= $memberAttributes{$A_USER_LAST_NAME} if $memberAttributes{$A_USER_LAST_NAME};
-			}
-	
-			# add user to contact role  (for VOs xml)
-			if(defined $resourcesAttributes{$A_R_APPDB_CONTACT_ROLE}) {
-				$voContacts->{$voShortName}->{$resourcesAttributes{$A_R_APPDB_CONTACT_ROLE} . "-" . $memberAttributes{$A_USER_ID}} = {
-						"name" => [ $userName ],
-						"role" => [ $resourcesAttributes{$A_R_APPDB_CONTACT_ROLE} ],
-						"email" => [ $memberAttributes{$A_USER_MAIL} ],
-						"sso" => [ {} ],
-						"eppn" => $memberAttributes{$A_USER_EPPNS},
-						"dn" => \@dn,
-				}
-			}
+		my @dn = ();
+		foreach my $subjectDN (keys %{$userCertDNs}) {
+			my $subjectDNWithoutPrefix = $subjectDN;
+			$subjectDNWithoutPrefix =~ s/^[0-9]+[:]//;
+			push @dn, { "content" => $subjectDNWithoutPrefix,
+				"ca" => $userCertDNs->{$subjectDN},
+			};
+		}
+
+		# prepare name
+		my $firstName = $data->getUserAttributeValue(member => $memberId, attrName => $A_USER_FIRST_NAME);
+		my $lastName = $data->getUserAttributeValue(member => $memberId, attrName => $A_USER_LAST_NAME);
 
-			#skip user if he was already processed for this VO
-			next if $userVoMap->{$memberAttributes{$A_USER_ID}}->{$vosAttributes{$A_VO_SHORT_NAME}}++;
+		my $userName = "";
+		if($firstName) {
+			$userName = $firstName;
+			$userName.= " " . $lastName if $lastName;
+		} else {
+			$userName.= $lastName if $lastName;
+		}
 
-			#prepare user record  (for users xml)
-			my @row = {
-									"uservo" => [ $userName ],
-									"vo" => [ $vosAttributes{$A_VO_SHORT_NAME} ],
-									"email" => [ $memberAttributes{$A_USER_MAIL} ],
-									"sso" => [ {} ],
-									"last_update" => [ "" ],
-									"first_update" => [ "" ],
-									"eppn" => $memberAttributes{$A_USER_EPPNS},
-									"dn" => \@dn,
-			          };
+		my $appDbContactRole = $data->getResourceAttributeValue(resource => $resourceId, attrName => $A_R_APPDB_CONTACT_ROLE);
 
+		my $voId = $data->getVoIdForResource(resource => $resourceId);
+		my $voShortName = $data->getVoAttributeValue(vo => $voId, attrName => $A_VO_SHORT_NAME);
 
-			#add user to global structure
-			push @{$usersXml->{"result"}}, { "row" => \@row };
+		my $userId = $data->getUserAttributeValue(member => $memberId, attrName => $A_USER_ID);
```

Comment:
This could be done by:
```
$data->getUserIdForMember( member => $memberId );
```

Reply:
You are right, I fixed it. We can generally remove all ``entity:core:id`` attributes from required once each service is converted to new hashed data structures.